### ISSUE TITLE
A question about the code

### ISSUE THREAD

**Issue by captainst** [JHart96/keras_ordinal_categorical_crossentropy#2]
I have some double on this line: `weights = K.cast(K.abs(K.argmax(y_true, axis=1) - K.argmax(y_pred, axis=1))/(K.int_shape(y_pred)[1] - 1), dtype='float32')`

Asseume that we have only 1 sample, 3 classes. 
Ground truth: [0, 0, 1]
Pred1: [0.4, 0.1, 0.5] 
Pred2: [0.1, 0.4, 0.5]

Then, for Ground truth, K.argmax(y_true, axis=1) == 2, and for both preds, K.argmax(y_pred, axis=1) == 2. So the weights == 0. Then there is no effect to the original cross-entropy loss.

But, Pred2 seems much better than Pred1.

**Comment by JHart96** [comment]
Apologies for the long delay in replying, it's been a busy time!

Yes I agree in principle that the penalties should be different between those two examples, but it gets tricky when trying to compute the loss according to the distribution of softmax scores over the categories. If instead we computed the sum of losses for each category weighted by the distance from the true category, the penalties will always be concentrated in such a way that the loss function will encourage values towards the middle-most category. This might be desired in some cases, since it provides the most conservative possible loss function, but this function was intended to be straightforward and without those biases, so I used the distance from the maximal score as the loss.

If you're looking for something that encourages a more conservative classifier, perhaps a loss function based on the total weighted distance from the true category would be the way to go.

I'm open to further suggestions, so if you have any ideas on how to improve this, let me know!

Thanks,
Jordan
